File: src/main/kotlin/day02/solution.kt
package day02

import java.io.File
import kotlin.math.abs

fun parse(inp: String) = inp.lines()
    .map { line -> line.split(" ").map { it.toInt() } }

fun safe(levels: List<Int>): Boolean {
    return levels in setOf(levels.sorted(), levels.sortedDescending())
            && levels.zipWithNext().all { (a, b) -> abs(a - b) in 1..3 }
}

fun first(inp: String) = parse(inp)
    .filter { safe(it) }
    .size

fun generate(levels: List<Int>) = levels.indices
    .map { levels.take(it) + levels.drop(it + 1) }

fun second(inp: String) = parse(inp)
    .filter { level -> generate(level).any { safe(it) } }
    .size

fun main() {
    val testInput = File("src/main/kotlin/day02/test-input.txt").readText()
    val input = File("src/main/kotlin/day02/input.txt").readText()
    println(first(input))
    println(second(input))
}

File: haskell/day02/solution.hs
import Data.List (sort, sortBy)
import Data.List.Split (splitOn)

parse :: String -> [[Int]]
parse inp = map (map read . splitOn " ") (lines inp)

safe :: [Int] -> Bool
safe xs =
  xs `elem` [sort xs, sortBy (flip compare) xs]
    && all
      (\x -> x `elem` [1 .. 3])
      (zipWith (\a b -> abs $ a - b) xs (tail xs))

first :: String -> Int
first inp = length $ filter safe (parse inp)

generate :: [Int] -> [[Int]]
generate xs = [take i xs ++ drop (i + 1) xs | i <- [0 .. (length xs)]]

second :: String -> Int
second inp = length $ filter (any safe . generate) (parse inp)

main :: IO ()
main = do
  testInput <- readFile "../../src/main/resources/day02/test-input.txt"
  input <- readFile "../../src/main/resources/day02/input.txt"
  print $ first input
  print $ second input
